Bug: too many arguments to function call counts.increment()

kemicoco opened the pull request:
Reproduce:
root@kemi-desktop:/home/kemi/git/kemi_bcc/bcc# ./tools/wakeuptime.py -f 3
/virtual/main.c:54:27: error: too many arguments to function call, expected
1, have 2
counts.increment(key, delta);
~~~~~~~~~~~~~~~~      ^~~~~
1 error generated.
Traceback (most recent call last):
  File "./tools/wakeuptime.py", line 165, in <module>
      b = BPF(text=bpf_text)
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py",
	  line 318, in __init__
  raise Exception("Failed to compile BPF text")
Exception: Failed to compile BPF text

root@kemi-desktop:/home/kemi/git/bcc# ./tools/offcputime.py -f 10
/virtual/main.c:55:27: error: too many arguments to function call, expected
1, have 2
counts.increment(key, delta);
~~~~~~~~~~~~~~~~      ^~~~~
1 error generated.
Traceback (most recent call last):
  File "./tools/offcputime.py", line 235, in <module>
     b = BPF(text=bpf_text)
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py",
 	line 318, in __init__
  raise Exception("Failed to compile BPF text")
Exception: Failed to compile BPF text

Signed-off-by: Kemi Wang <kemi.wang@intel.com>

kemicoco commented:
On 2018年08月20日 16:48, Paul Chaignon wrote:
> pchaigno requested changes on this pull request.
> 
> How did you install bcc? 
I used ubuntu 14.04 with 4.18+ kernel

First: build based-on the latest linus tree.
a) enable the config option according to INSTALL.md
b) make && make modules_install && make headers_install && make install
c) reboot

Second: update repo
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys D4284CDD     
echo "deb https://repo.iovisor.org/apt/xenial xenial main" | sudo tee /etc/apt/sources.list.d/iovisor.list
sudo apt-get update                                                        
sudo apt-get install bcc-tools libbcc-examples   // I remove linux-header-$(uname -r) since I have run "make headers_install" above


Third: install dependency
# All versions                                                                
sudo apt-get -y install bison build-essential cmake flex git libedit-dev \ 
  libllvm3.7 llvm-3.7-dev libclang-3.7-dev python zlib1g-dev libelf-dev       
                                                                              
# For Lua support                                                             
sudo apt-get -y install luajit luajit-5.1-dev


Forth: build bcc
### Install and compile BCC                                                
```                                                                        
git clone https://github.com/iovisor/bcc.git                               
mkdir bcc/build; cd bcc/build                                              
cmake .. -DCMAKE_INSTALL_PREFIX=/home/kemi/git/bcc/usr    //I changed the install directory                                     
make                                                                       
sudo make install

> Are you sure the installed version of bcc matches the version of the tools you use? 

I am not sure. Anything you want to me to check?

> The optional, additional parameter for `increment` was added in #1897 so the tools look correct.

When I search the function call counts.increment() under bcc/ directory via "egrep counts.increment -r -n *"
I found only tools/offcputime.py and tools/offwaketime.py use two parameters for function counts.increment().
So, I removed the parameter "delta " from counts.increment function call. And, it works and I got the CPU 
frame graph finally with the help of Gregg's tool.

> 
> 
>

pchaigno commented:
> I am not sure. Anything you want to me to check?

You installed bcc from the packages (last release), but used the tools from the repository. So the versions don't match. If you either install bcc from sources or used the tools as installed with bcc from packages, it should work.

> So, I removed the parameter "delta " from counts.increment function call. And, it works and I got the CPU frame graph finally with the help of Gregg's tool.

Yes, but that's not the correct fix. It's only required in your case because the versions of bcc and its tools don't match on your system.

kemicoco commented:
On 2018年08月20日 17:51, Paul Chaignon wrote:
>> I am not sure. Anything you want to me to check?
> 
> You installed bcc from the packages (last release), but used the tools from the repository. So the versions don't match. If you either install bcc from sources or used the tools as installed with bcc from packages, it should work.
> 

I want to try the bcc installation from source, and check its effectiveness.
So, I removed the bcc-tools and libbcc-examples via "apt-get remove bcc-tools libbcc-examples"

And recompile bcc from the source. It still does not work unless changing the function call counts.increment().

recompile bcc:
cd build && make clean 
cd ../
rm -rf usr/ build/

### Install and compile BCC                                                   
```                                                                                                          
mkdir bcc/build; cd bcc/build                                                 
cmake .. -DCMAKE_INSTALL_PREFIX=/home/kemi/git/bcc/usr                                          
make                                                                          
sudo make install                                                             
```   

>> So, I removed the parameter "delta " from counts.increment function call. And, it works and I got the CPU frame graph finally with the help of Gregg's tool.
> 
> Yes, but that's not the correct fix. It's only required in your case because the versions of bcc and its tools don't match on your system.
>

pchaigno commented:
Can you check the bcc module imported by the script is actually the one you installed from sources?

kemicoco commented:
To make sure I used the bcc from source, I removed bcc from /usr/lib/python2.7 via “apt-get remove python-bcc” first.
Then, run ./tool/offcputime.py, it complains “ImportError: No module named bcc “

Next, I added the bcc module search path  to the PYTHONPATH. And rerun ./tool/offcputime.py.
See the testing result below.

root@kemi-desktop:/home/kemi/git/bcc# find /usr/lib -name bcc
root@kemi-desktop:/home/kemi/git/bcc#
root@kemi-desktop:/home/kemi/git/bcc# export PYTHONPATH=/home/kemi/git/bcc/usr/lib/python2.7/dist-packages
root@kemi-desktop:/home/kemi/git/bcc#
root@kemi-desktop:/home/kemi/git/bcc# python
Python 2.7.12 (default, Nov 20 2017, 18:23:56)
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.path
['', '/home/kemi/git/bcc/usr/lib/python2.7/dist-packages', '/usr/lib/python2.7', '/usr/lib/python2.7/plat-x86_64-linux-gnu', '/usr/lib/python2.7/lib-tk', '/usr/lib/python2.7/lib-old', '/usr/lib/python2.7/lib-dynload', '/usr/local/lib/python2.7/dist-packages', '/usr/lib/python2.7/dist-packages', '/usr/lib/python2.7/dist-packages/gtk-2.0', '/usr/lib/python2.7/dist-packages/wx-3.0-gtk2']
>>>
>>>
root@kemi-desktop:/home/kemi/git/bcc#
root@kemi-desktop:/home/kemi/git/bcc#
root@kemi-desktop:/home/kemi/git/bcc# ./tools/offcputime.py
/virtual/main.c:55:27: error: too many arguments to function call, expected 1, have 2
    counts.increment(key, delta);
    ~~~~~~~~~~~~~~~~      ^~~~~
1 error generated.
Traceback (most recent call last):
  File "./tools/offcputime.py", line 235, in <module>
    b = BPF(text=bpf_text)
  File "/home/kemi/git/bcc/usr/lib/python2.7/dist-packages/bcc/__init__.py", line 318, in __init__
    raise Exception("Failed to compile BPF text")
Exception: Failed to compile BPF text
root@kemi-desktop:/home/kemi/git/bcc#



From: Paul Chaignon [mailto:notifications@github.com]
Sent: Monday, August 20, 2018 6:43 PM
To: iovisor/bcc <bcc@noreply.github.com>
Cc: Wang, Kemi <kemi.wang@intel.com>; Author <author@noreply.github.com>
Subject: Re: [iovisor/bcc] Bug: too many arguments to function call counts.increment() (#1937)


Can you check the bcc module imported by the script is actually the one you installed from sources?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/iovisor/bcc/pull/1937#issuecomment-414273961>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AcDuwyr4WEwZNYAG0yaAJKKuYtzK7HZwks5uSpLEgaJpZM4WDk0K>.

pchaigno commented:
I don't have any other ideas; maybe someone else will. It works in the Jenkins builds and on my own system, so I'm guessing it's an issue with your configuration, but I could be wrong.

yonghong-song commented:
Could you check whether your source code bcc:src/cc/frontends/clang/b_frontend_action.cc having the following codes
```
        } else if (memb_name == "increment") {
          string name = Ref->getDecl()->getName();
          string arg0 = rewriter_.getRewrittenText(expansionRange(Call->getArg(0)->getSourceRange()));

          string increment_value = "1";
          if (Call->getNumArgs() == 2) {
            increment_value = rewriter_.getRewrittenText(expansionRange(Call->getArg(1)->getSourceRange()));

          }

          string lookup = "bpf_map_lookup_elem_(bpf_pseudo_fd(1, " + fd + ")";
          string update = "bpf_map_update_elem_(bpf_pseudo_fd(1, " + fd + ")";
          txt  = "({ typeof(" + name + ".key) _key = " + arg0 + "; ";
          txt += "typeof(" + name + ".leaf) *_leaf = " + lookup + ", &_key); ";

          txt += "if (_leaf) (*_leaf) += " + increment_value + ";";
          if (desc->second.type == BPF_MAP_TYPE_HASH) {
            txt += "else { typeof(" + name + ".leaf) _zleaf; __builtin_memset(&_zleaf, 0, sizeof(_zleaf)); ";
            txt += "_zleaf += " + increment_value + ";";
            txt += update + ", &_key, &_zleaf, BPF_NOEXIST); } ";
          }
          txt += "})";
```
I am wondering whether this is due to a mismatch between C++ code and tools code for a particular checkout or package?

kemicoco commented:
On 2018年08月21日 00:53, yonghong-song wrote:
> Could you check whether your source code bcc:src/cc/frontends/clang/b_frontend_action.cc having the following codes
> ```
>         } else if (memb_name == "increment") {
>           string name = Ref->getDecl()->getName();
>           string arg0 = rewriter_.getRewrittenText(expansionRange(Call->getArg(0)->getSourceRange()));
> 
>           string increment_value = "1";
>           if (Call->getNumArgs() == 2) {
>             increment_value = rewriter_.getRewrittenText(expansionRange(Call->getArg(1)->getSourceRange()));
> 
>           }
> 
>           string lookup = "bpf_map_lookup_elem_(bpf_pseudo_fd(1, " + fd + ")";
>           string update = "bpf_map_update_elem_(bpf_pseudo_fd(1, " + fd + ")";
>           txt  = "({ typeof(" + name + ".key) _key = " + arg0 + "; ";
>           txt += "typeof(" + name + ".leaf) *_leaf = " + lookup + ", &_key); ";
> 
>           txt += "if (_leaf) (*_leaf) += " + increment_value + ";";
>           if (desc->second.type == BPF_MAP_TYPE_HASH) {
>             txt += "else { typeof(" + name + ".leaf) _zleaf; __builtin_memset(&_zleaf, 0, sizeof(_zleaf)); ";
>             txt += "_zleaf += " + increment_value + ";";
>             txt += update + ", &_key, &_zleaf, BPF_NOEXIST); } ";
>           }
>           txt += "})";
> ```

Yes. It has.
My bcc source code is got from master tree.

780         } else if (memb_name == "increment") {                             
 781           string name = Ref->getDecl()->getName();                         
 782           string arg0 = rewriter_.getRewrittenText(expansionRange(Call->getArg(0)->getSourceRange()));
 783                                                                            
 784           string increment_value = "1";                                    
 785           if (Call->getNumArgs() == 2) {                                   
 786             increment_value = rewriter_.getRewrittenText(expansionRange(Call->getArg(1)->getSourceRange()));
 787                                                                            
 788           }                                                                
 789                                                                            
 790           string lookup = "bpf_map_lookup_elem_(bpf_pseudo_fd(1, " + fd + ")";
 791           string update = "bpf_map_update_elem_(bpf_pseudo_fd(1, " + fd + ")";
 792           txt  = "({ typeof(" + name + ".key) _key = " + arg0 + "; ";      
 793           txt += "typeof(" + name + ".leaf) *_leaf = " + lookup + ", &_key); ";
 794                                                                            
 795           txt += "if (_leaf) (*_leaf) += " + increment_value + ";";        
 796           if (desc->second.type == BPF_MAP_TYPE_HASH) {                    
 797             txt += "else { typeof(" + name + ".leaf) _zleaf; __builtin_memset(&_zleaf, 0, sizeof(_zleaf)); ";
 798             txt += "_zleaf += " + increment_value + ";";                   
 799             txt += update + ", &_key, &_zleaf, BPF_NOEXIST); } ";          
 800           }                                                                
 801           txt += "})";                                                     
 802         } else if (memb_name == "perf_submit") {                           
 803           string name = Ref->getDecl()->getName();                         
 804           string arg0 = rewriter_.getRewrittenText(expansionRange(Call->getArg(0)->getSourceRange()));
 805           string args_other = rewriter_.getRewrittenText(expansionRange(SourceRange(Call->getArg(1)->getLocStart(),
 806                                                            Call->getArg(2)->getLocEnd())));
 807           txt = "bpf_perf_event_output(" + arg0 + ", bpf_pseudo_fd(1, " + fd + ")";          

> I am wondering whether this is due to a mismatch between C++ code and tools code for a particular checkout or package?
> 

I have no idea for that. But I got some warning during executing "make" command, not sure whether it is the root cause or not.
Please check the attachment(make.log) for that. Maybe you can find some clues.

palmtenor commented:
Can you try a newer version of LLVM? The oldest LLVM version we have on Jenkens is 3.8.0 (Ubuntu 1604). Maybe 3.7 doesn't compile the `(, ...)` syntax correctly?

kemicoco commented:
On 2018年08月21日 16:13, Teng Qin wrote:
> Can you try a newer version of LLVM? The oldest LLVM version we have on Jenkens is 3.8.0 (Ubuntu 1604). Maybe 3.7 doesn't compile the `(, ...)` syntax correctly?
> 
> 

First: apt-get remove libllvm3.7 llvm-3.7-dev

Second: apt-get install libllvm3.8 llvm-3.8-dev

Third: compile bcc
root@kemi-desktop:/home/kemi/git/bcc/build# cmake .. -DCMAKE_INSTALL_PREFIX=/home/kemi/git/bcc/usr
-- Latest recognized Git tag is v0.6.1
-- Git HEAD is 45b6babd74dbfe81ead467e10c83c8712c011aa4
-- Revision is 0.6.1-45b6babd
CMake Error at /usr/share/llvm/cmake/LLVMExports.cmake:1034 (message):
  The imported target "LLVMSupport" references the file

     "/usr/lib/libLLVMSupport.a"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "/usr/share/llvm/cmake/LLVMExports.cmake"

  but not all the files it references.

Call Stack (most recent call first):
  /usr/share/llvm/cmake/LLVMConfig.cmake:178 (include)
  CMakeLists.txt:29 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/kemi/git/bcc/build/CMakeFiles/CMakeOutput.log".
root@kemi-desktop:/home/kemi/git/bcc/build# 



It seems a knowns bug for llvm-3.8.
See this bug report: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=819072

palmtenor commented:
OK, could you try a newer version then? BCC is developed against latest LLVM version and either 3.7 or 3.8 are pretty old now.

kemicoco commented:
On 2018年08月22日 14:21, yonghong-song wrote:
> This should not be a 3.7 llvm compiler issue. We have 3.7 compiler in ubuntu 16.04 testbot and it works fine. do you want to add some instrumentation in the above bcc code to check what is the real problem or the code is called at all?
> 

I　would like to dig it out if you show me some suggestion on how to do that:)

yonghong-song commented:
@kemicoco just throw some printf in that piece of code and see whether it is hit or performs as expected.
